crates/sui-kvstore/src/writer.rs

```diff
@@ -0,0 +1,118 @@
+                "Latest checkpoint to have been uploaded to the remote key value store",
+                registry
+            )
+            .unwrap(),
+        }
+    }
+}
+
+pub async fn setup_key_value_store_uploader(
+    store: RocksDbStore,
+    config: &Option<KVStoreConfig>,
+    registry: &Registry,
+) -> Result<Option<oneshot::Sender<()>>> {
+    if config.is_none() {
+        return Ok(None);
+    }
+    let client = DynamoDbClient::new(config.as_ref().unwrap()).await;
+    let checkpoint_number = client
+        .get_state()
+        .await
+        .expect("failed to fetch key value uploader state")
+        .unwrap_or_default();
+    let (sender, receiver) = oneshot::channel();
+    info!(
+        "Starting backfilling kev value store. Current checkpoint is {}",
+        checkpoint_number
+    );
+
+    let metrics = KVStoreMetrics::new(registry);
+    spawn_monitored_task!(async move {
+        upload_to_kv_store(store, client, checkpoint_number, receiver, metrics)
+            .await
+            .expect("Upload failer to key value store")
+    });
+    Ok(Some(sender))
+}
+
+async fn upload_to_kv_store<S, C>(
+    store: S,
+    mut client: C,
+    mut checkpoint_number: CheckpointSequenceNumber,
+    mut receiver: oneshot::Receiver<()>,
+    metrics: KVStoreMetrics,
+) -> Result<()>
+where
+    S: ReadStore,
+    C: KVWriteClient,
+    <S as ReadStore>::Error: Send,
+{
+    while receiver.try_recv().is_err() {
+        match store
+            .get_full_checkpoint_contents_by_sequence_number(checkpoint_number + 1)
+            .map_err(|_| anyhow!("Failed to read checkpoint contents from the store"))?
+        {
+            Some(contents) => {
+                let mut transactions = vec![];
+                let mut effects = vec![];
+                let mut events = vec![];
+
+                for content in contents.iter() {
+                    let transaction_digest = content.transaction.digest().into_inner().to_vec();
+                    effects.push((transaction_digest.clone(), content.effects.clone()));
+                    transactions.push((transaction_digest, content.transaction.clone()));
+
+                    if let Some(event_digest) = content.effects.events_digest() {
+                        if let Some(tx_events) = store
+                            // TODO: check events sequence
+                            .get_transaction_events(event_digest)
+                            .map_err(|_| anyhow!("Failed to fetch events from the store"))?
+                        {
+                            events.push((event_digest.into_inner().to_vec(), tx_events));
+                        }
+                    }
+                }
+                client
+                    .multi_set(KVTable::Transactions, transactions)
+                    .await?;
+                client.multi_set(KVTable::Effects, effects).await?;
+                client.multi_set(KVTable::Events, events).await?;
```

Comment:
do you think its worth doing the multisets concurrently here, or is it already fast enough?

Reply:
let's try vanilla version first and iterate on top
we anyway will have to deploy it to one of the hosts manually not to start the backfill process